src/jni_core/src/jvm_runtime.rs

```diff
@@ -16,86 +16,106 @@ use core::option::Option::Some;
-        }
-    } else {
-        return Err(ErrorCode::InternalError(format!(
-            "failed to read CONNECTOR_LIBS_PATH \"{}\"",
-            libs_path
-        ))
-        .into());
+    pub fn get_or_init(&self) -> Result<&JavaVM, &RwError> {
+        self.0.get_or_init(Self::inner_new).as_ref()
     }
 
-    let jvm_heap_size = if let Ok(heap_size) = std::env::var("JVM_HEAP_SIZE") {
-        heap_size
-    } else {
-        // Use 10% of total memory by default
-        // TODO: should use compute-node's total_memory_bytes
-        format!("{}", system_memory_available_bytes() / 10)
-    };
-
-    // Build the VM properties
-    let args_builder = InitArgsBuilder::new()
-        // Pass the JNI API version (default is 8)
-        .version(JNIVersion::V8)
-        .option("-ea")
-        .option("-Dis_embedded_connector=true")
-        .option(format!("-Djava.class.path={}", class_vec.join(":")))
-        .option(format!("-Xmx{}", jvm_heap_size));
-
-    tracing::info!("JVM args: {:?}", args_builder);
-    let jvm_args = args_builder.build().unwrap();
-
-    // Create a new VM
-    let jvm = match JavaVM::new(jvm_args) {
-        Err(err) => {
-            tracing::error!("fail to new JVM {:?}", err);
-            return Err(ErrorCode::InternalError("fail to new JVM".to_string()).into());
+    fn inner_new() -> Result<JavaVM, RwError> {
+        let libs_path = if let Ok(libs_path) = std::env::var("CONNECTOR_LIBS_PATH") {
+            libs_path
+        } else {
+            return Err(ErrorCode::InternalError(
+                "environment variable CONNECTOR_LIBS_PATH is not specified".to_string(),
+            )
+            .into());
+        };
+
+        let dir = Path::new(&libs_path);
+
+        if !dir.is_dir() {
+            return Err(ErrorCode::InternalError(format!(
+                "CONNECTOR_LIBS_PATH \"{}\" is not a directory",
+                libs_path
+            ))
+            .into());
         }
-        Ok(jvm) => jvm,
-    };
 
-    tracing::info!("initialize JVM successfully");
+        let mut class_vec = vec![];
 
-    register_native_method_for_jvm(&jvm).unwrap();
+        if let Ok(entries) = fs::read_dir(dir) {
+            for entry in entries.flatten() {
+                let entry_path = entry.path();
+                if entry_path.file_name().is_some() {
+                    let path = std::fs::canonicalize(entry_path)?;
+                    class_vec.push(path.to_str().unwrap().to_string());
+                }
+            }
+        } else {
+            return Err(ErrorCode::InternalError(format!(
+                "failed to read CONNECTOR_LIBS_PATH \"{}\"",
+                libs_path
+            ))
+            .into());
+        }
 
-    Ok(jvm)
-});
+        let jvm_heap_size = if let Ok(heap_size) = std::env::var("JVM_HEAP_SIZE") {
+            heap_size
+        } else {
+            // Use 10% of total memory by default
+            // TODO: should use compute-node's total_memory_bytes
```

Comment:
Oh this TODO is still here. Need to be fixed 🥹

Reply:
How about using `system_memory_available_bytes() as f64 * DEFAULT_MEMORY_PROPORTION`?

Reply:
It seems we still need to rely on the cloud to specify for us. cc @arkbriar ark